### PR TITLE
Cmake presets

### PR DESCRIPTION
**Pull request by cmcfarlen**
CMakePresets.json will let you configure the project with a known set of options.  This can be useful for reproducing builds and should give us more outside control over CI pipeline builds.

To configure with a preset just use the `--preset` option to cmake and provide the name.  For example, to build `autests`:

```
cmake . --preset autest
cmake --build --preset autest
```

You can provide a `CMakeUserPresets.json` file to add additional presets.  This file is not to be checked in, and is included in `.gitignore`.  Here is an example:

```json
{
  "version": 2,
  "cmakeMinimumRequired": {
    "major": 3,
    "minor": 20,
    "patch": 0
  },
  "configurePresets": [
    {
      "name": "my-dev",
      "displayName": "my development",
      "description": "My Development Presets",
      "inherits": ["dev"],
      "cacheVariables": {
        "CMAKE_INSTALL_PREFIX": "/opt/ats-cmake",
        "jemalloc_ROOT": "/opt/homebrew"
      }
    }
  ]
}
```

This is like the `dev` preset, but changes the install prefix and the jemalloc root.

This is just to get us started.  As we get closer to using cmake, I'm sure we will expand this.

### CODE REVIEW

**Comment by cmcfarlen** [comment]
[approve ci autest]

**Review by bneradt — CHANGES_REQUESTED** [review]
Look great.

I suggest adding `CMAKE_COMPILE_WARNING_AS_ERROR` to more configurations. Devs find it annoying if they use one of these locally and find in CI that they had some warnings they missed because they weren't treated as errors.

**Comment by cmcfarlen** [comment]
> Look great.
> 
> I suggest adding `CMAKE_COMPILE_WARNING_AS_ERROR` to more configurations. Devs find it annoying if they use one of these locally and find in CI that they had some warnings they missed because they weren't treated as errors.

With the inheritance, every preset should have this set to `ON`.  The top level ones are `default` and `ci`

**Comment by bneradt** [comment]
> > Look great.
> > I suggest adding `CMAKE_COMPILE_WARNING_AS_ERROR` to more configurations. Devs find it annoying if they use one of these locally and find in CI that they had some warnings they missed because they weren't treated as errors.
> 
> With the inheritance, every preset should have this set to `ON`. The top level ones are `default` and `ci`

Ah, great. Thanks.

File: CMakePresets.json [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    },
+    {
+      "name": "dev",
+      "displayName": "development",
+      "description": "Development Presets",
+      "inherits": ["default"],
+      "binaryDir": "${sourceDir}/cmake-build-${presetName}",
+      "generator": "Ninja",
+      "cacheVariables": {
+        "CMAKE_EXPORT_COMPILE_COMMANDS": "ON",
```
Comment (bneradt) [inline CMakePresets.json:39]: Nice.

**Review by bneradt — APPROVED** [review]

**Pull request merged**